File: extension/core/review.js
// Retrospectiva semanal (spec/SPEC.md §4.7) — a revisão à la Safren: o que
// foi feito, quando, e o que interrompeu. A semana projetada (seg–dom) é a
// que contém a data local de `now`; navegar no histórico é variar o `now`.

import { normalize } from "./envelope.js";
import { localDate, addDays } from "./time.js";

// mondayOf: segunda-feira da semana que contém a data civil dada.
function mondayOf(isoDate) {
  const [y, m, d] = isoDate.split("-").map(Number);
  const weekday = new Date(Date.UTC(y, m - 1, d)).getUTCDay(); // 0 = domingo
  return addDays(isoDate, -((weekday + 6) % 7));
}

// Duração em minutos (arredondada); null para intervalos vazios/invertidos.
function minutesBetween(fromTs, toTs) {
  const ms = Date.parse(toTs) - Date.parse(fromTs);
  if (!Number.isFinite(ms) || ms <= 0) return null;
  return Math.round(ms / 60000);
}

export function project(events, { now, tz }) {
  const weekStart = mondayOf(localDate(now, tz));
  const weekEnd = addDays(weekStart, 6);

  const days = {};
  for (let i = 0; i < 7; i++) {
    days[addDays(weekStart, i)] = {
      reviews: 0, focus_minutes: 0, pause_minutes: 0, completed: 0, interrupted: 0,
      checkins: 0, checkins_on_task: 0,
    };
  }
  const inWeek = (day) => day >= weekStart && day <= weekEnd;

  const sessions = new Map();
  const pauses = new Map();
  const ensure = (map, id) => {
    if (!map.has(id)) map.set(id, {});
    return map.get(id);
  };

  const reasons = {};
  let appointment_minutes = 0;

  for (const event of normalize(events)) {
    const { payload } = event;
    if (event.type === "session.started") {
      ensure(sessions, payload.session_id).startTs = event.ts;
    } else if (event.type === "session.ended") {
      const s = ensure(sessions, payload.session_id);
      s.endTs = event.ts;
      s.outcome = payload.outcome;
      if (payload.reason !== undefined) s.reason = payload.reason; // v2; v1 ≡ sem motivo
    } else if (event.type === "pause.started") {
      ensure(pauses, payload.pause_id).startTs = event.ts;
    } else if (event.type === "pause.ended") {
      ensure(pauses, payload.pause_id).endTs = event.ts;
    } else if (event.type === "pause.logged") {
      // Retroativa: o tempo do domínio vem no payload, não no ts.
      const p = ensure(pauses, payload.pause_id);
      p.startTs = payload.starts_at;
      p.endTs = payload.ends_at;
    } else if (event.type === "card.reviewed") {
      const day = localDate(event.ts, tz);
      if (inWeek(day)) days[day].reviews += 1;
    } else if (event.type === "checkin.logged") {
      // Automonitoramento de atenção: "na tarefa em X de Y checagens".
      const day = localDate(event.ts, tz);
      if (inWeek(day)) {
        days[day].checkins += 1;
        if (payload.on_task) days[day].checkins_on_task += 1;
      }
    } else if (event.type === "appointment.created") {
      const minutes = minutesBetween(payload.starts_at, payload.ends_at);
      const day = localDate(payload.starts_at, tz);
      if (minutes !== null && inWeek(day)) appointment_minutes += minutes;
    }
  }

  for (const s of sessions.values()) {
    if (!s.startTs || !s.endTs) continue; // par não fechado é ignorado
    const minutes = minutesBetween(s.startTs, s.endTs);
    if (minutes === null) continue;
    const day = localDate(s.startTs, tz);
    if (!inWeek(day)) continue;
    days[day].focus_minutes += minutes;
    if (s.outcome === "interrupted") {
      days[day].interrupted += 1;
      if (s.reason) reasons[s.reason] = (reasons[s.reason] ?? 0) + 1;
    } else {
      days[day].completed += 1;
    }
  }

  for (const p of pauses.values()) {
    if (!p.startTs || !p.endTs) continue;
    const minutes = minutesBetween(p.startTs, p.endTs);
    if (minutes === null) continue;
    const day = localDate(p.startTs, tz);
    if (inWeek(day)) days[day].pause_minutes += minutes;
  }

  const totals = {
    reviews: 0, focus_minutes: 0, pause_minutes: 0, completed: 0, interrupted: 0,
    checkins: 0, checkins_on_task: 0,
  };
  for (const d of Object.values(days)) {
    totals.reviews += d.reviews;
    totals.focus_minutes += d.focus_minutes;
    totals.pause_minutes += d.pause_minutes;
    totals.completed += d.completed;
    totals.interrupted += d.interrupted;
    totals.checkins += d.checkins;
    totals.checkins_on_task += d.checkins_on_task;
  }

  return { week_start: weekStart, week_end: weekEnd, days, totals, reasons, appointment_minutes };
}

File: cli/core/review.go
// Retrospectiva semanal (spec/SPEC.md §4.7) — a revisão à la Safren: o que
// foi feito, quando, e o que interrompeu. A semana projetada (seg–dom) é a
// que contém a data local de `now`; navegar no histórico é variar o `now`.
package core

import (
	"math"
	"time"
)

type ReviewDay struct {
	Reviews        int `json:"reviews"`
	FocusMinutes   int `json:"focus_minutes"`
	PauseMinutes   int `json:"pause_minutes"`
	Completed      int `json:"completed"`
	Interrupted    int `json:"interrupted"`
	Checkins       int `json:"checkins"`
	CheckinsOnTask int `json:"checkins_on_task"`
}

type ReviewState struct {
	WeekStart          string               `json:"week_start"`
	WeekEnd            string               `json:"week_end"`
	Days               map[string]ReviewDay `json:"days"`
	Totals             ReviewDay            `json:"totals"`
	Reasons            map[string]int       `json:"reasons"`
	AppointmentMinutes int                  `json:"appointment_minutes"`
}

func init() {
	Register("review", func(events []Event, p Params) (any, error) {
		return Review(events, p)
	})
}

func Review(events []Event, p Params) (ReviewState, error) {
	today, err := LocalDate(p.Now, p.TZ)
	if err != nil {
		return ReviewState{}, err
	}
	weekStart, err := mondayOf(today)
	if err != nil {
		return ReviewState{}, err
	}
	weekEnd, err := AddDays(weekStart, 6)
	if err != nil {
		return ReviewState{}, err
	}

	state := ReviewState{
		WeekStart: weekStart,
		WeekEnd:   weekEnd,
		Days:      make(map[string]ReviewDay, 7),
		Reasons:   map[string]int{},
	}
	for i := 0; i < 7; i++ {
		day, err := AddDays(weekStart, i)
		if err != nil {
			return ReviewState{}, err
		}
		state.Days[day] = ReviewDay{}
	}
	inWeek := func(day string) bool { return day >= weekStart && day <= weekEnd }

	type interval struct {
		startTS, endTS  string
		outcome, reason string
	}
	sessions := map[string]*interval{}
	pauses := map[string]*interval{}
	ensure := func(m map[string]*interval, id string) *interval {
		if m[id] == nil {
			m[id] = &interval{}
		}
		return m[id]
	}

	for _, event := range Normalize(events) {
		switch event.Type {
		case "session.started":
			pl, err := decodePayload[struct {
				SessionID string `json:"session_id"`
			}](event)
			if err != nil {
				return ReviewState{}, err
			}
			ensure(sessions, pl.SessionID).startTS = event.TS

		case "session.ended":
			pl, err := decodePayload[struct {
				SessionID string  `json:"session_id"`
				Outcome   string  `json:"outcome"`
				Reason    *string `json:"reason"` // v2; v1 ≡ sem motivo
			}](event)
			if err != nil {
				return ReviewState{}, err
			}
			s := ensure(sessions, pl.SessionID)
			s.endTS = event.TS
			s.outcome = pl.Outcome
			if pl.Reason != nil {
				s.reason = *pl.Reason
			}

		case "pause.started":
			pl, err := decodePayload[struct {
				PauseID string `json:"pause_id"`
			}](event)
			if err != nil {
				return ReviewState{}, err
			}
			ensure(pauses, pl.PauseID).startTS = event.TS

		case "pause.ended":
			pl, err := decodePayload[struct {
				PauseID string `json:"pause_id"`
			}](event)
			if err != nil {
				return ReviewState{}, err
			}
			ensure(pauses, pl.PauseID).endTS = event.TS

		case "pause.logged":
			// Retroativa: o tempo do domínio vem no payload, não no ts.
			pl, err := decodePayload[struct {
				PauseID  string `json:"pause_id"`
				StartsAt string `json:"starts_at"`
				EndsAt   string `json:"ends_at"`
			}](event)
			if err != nil {
				return ReviewState{}, err
			}
			p := ensure(pauses, pl.PauseID)
			p.startTS, p.endTS = pl.StartsAt, pl.EndsAt

		case "card.reviewed":
			day, err := LocalDate(event.TS, p.TZ)
			if err != nil {
				return ReviewState{}, err
			}
			if inWeek(day) {
				d := state.Days[day]
				d.Reviews++
				state.Days[day] = d
			}

		case "checkin.logged":
			// Automonitoramento de atenção: "na tarefa em X de Y checagens".
			pl, err := decodePayload[struct {
				OnTask bool `json:"on_task"`
			}](event)
			if err != nil {
				return ReviewState{}, err
			}
			day, err := LocalDate(event.TS, p.TZ)
			if err != nil {
				return ReviewState{}, err
			}
			if inWeek(day) {
				d := state.Days[day]
				d.Checkins++
				if pl.OnTask {
					d.CheckinsOnTask++
				}
				state.Days[day] = d
			}

		case "appointment.created":
			pl, err := decodePayload[struct {
				StartsAt string `json:"starts_at"`
				EndsAt   string `json:"ends_at"`
			}](event)
			if err != nil {
				return ReviewState{}, err
			}
			day, err := LocalDate(pl.StartsAt, p.TZ)
			if err != nil {
				return ReviewState{}, err
			}
			if minutes, ok := minutesBetween(pl.StartsAt, pl.EndsAt); ok && inWeek(day) {
				state.AppointmentMinutes += minutes
			}
		}
	}

	for _, s := range sessions {
		if s.startTS == "" || s.endTS == "" {
			continue // par não fechado é ignorado
		}
		minutes, ok := minutesBetween(s.startTS, s.endTS)
		if !ok {
			continue
		}
		day, err := LocalDate(s.startTS, p.TZ)
		if err != nil {
			return ReviewState{}, err
		}
		if !inWeek(day) {
			continue
		}
		d := state.Days[day]
		d.FocusMinutes += minutes
		if s.outcome == "interrupted" {
			d.Interrupted++
			if s.reason != "" {
				state.Reasons[s.reason]++
			}
		} else {
			d.Completed++
		}
		state.Days[day] = d
	}

	for _, pause := range pauses {
		if pause.startTS == "" || pause.endTS == "" {
			continue
		}
		minutes, ok := minutesBetween(pause.startTS, pause.endTS)
		if !ok {
			continue
		}
		day, err := LocalDate(pause.startTS, p.TZ)
		if err != nil {
			return ReviewState{}, err
		}
		if !inWeek(day) {
			continue
		}
		d := state.Days[day]
		d.PauseMinutes += minutes
		state.Days[day] = d
	}

	for _, d := range state.Days {
		state.Totals.Reviews += d.Reviews
		state.Totals.FocusMinutes += d.FocusMinutes
		state.Totals.PauseMinutes += d.PauseMinutes
		state.Totals.Completed += d.Completed
		state.Totals.Interrupted += d.Interrupted
		state.Totals.Checkins += d.Checkins
		state.Totals.CheckinsOnTask += d.CheckinsOnTask
	}
	return state, nil
}

// mondayOf devolve a segunda-feira da semana que contém a data civil dada.
func mondayOf(isoDate string) (string, error) {
	t, err := time.Parse(civilDate, isoDate)
	if err != nil {
		return "", err
	}
	return AddDays(isoDate, -((int(t.Weekday()) + 6) % 7))
}

// minutesBetween devolve a duração em minutos (arredondada) entre dois
// instantes; ok=false para intervalos vazios ou invertidos, que são ignorados.
func minutesBetween(fromTS, toTS string) (int, bool) {
	from, err := time.Parse(time.RFC3339, fromTS)
	if err != nil {
		return 0, false
	}
	to, err := time.Parse(time.RFC3339, toTS)
	if err != nil {
		return 0, false
	}
	d := to.Sub(from)
	if d <= 0 {
		return 0, false
	}
	return int(math.Round(d.Minutes())), true
}
